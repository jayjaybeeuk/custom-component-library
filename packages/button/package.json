{
  "name": "@jayjaybeeuk/button",
  "version": "1.0.0",
  "description": "",
  "main": "lib/button.js",
  "typings": "lib/button.d.ts",
  "directories": {
    "lib": "lib",
    "test": "__tests__"
  },
  "files": ["lib"],
  "publishConfig": {
    "access": "public"
  },
  "scripts": {
    "build": "tsc",
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "devDependencies": {
    "@types/react": "^18.3.3",
    "typescript": "^5.4.5"
  },
  "dependencies": {
    "@radix-ui/themes": "^3.0.5",
    "react": "^18.3.1"
  }
}
