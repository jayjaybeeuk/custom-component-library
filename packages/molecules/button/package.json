{
  "name": "@jayjaybeeuk/molecules-button",
  "version": "1.0.2",
  "description": "",
  "main": "lib/button.js",
  "type": "module",
  "typings": "lib/button.d.ts",
  "directories": {
    "lib": "lib",
    "test": "__tests__"
  },
  "files": [
    "lib"
  ],
  "publishConfig": {
    "access": "public"
  },
  "scripts": {
    "build": "../../../scripts/build.sh",
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "devDependencies": {
    "@types/react": "^18.3.3",
    "typescript": "^5.4.5"
  },
  "dependencies": {
    "@jayjaybeeuk/types": "^1.0.1",
    "@radix-ui/themes": "^3.0.5",
    "react": "^18.3.1"
  }
}
