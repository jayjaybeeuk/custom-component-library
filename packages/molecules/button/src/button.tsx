import { Icon } from '@jayjaybeeuk/atoms-icon';
import { StartEndIconProps } from '@jayjaybeeuk/types';
import { Button as RadixButton, Text } from '@radix-ui/themes';
import clsx from 'clsx';
import { forwardRef } from 'react';
import styles from './style.module.css';


export interface ButtonProps
  extends Omit<React.ButtonHTMLAttributes<HTMLButtonElement>, 'color'>,
    StartEndIconProps {
  variant?:
    | 'default'
    | 'positive'
    | 'negative'
    | 'grey'
    | 'grey-red'
    | 'dropdown'
    | 'ghost-default'
    | 'ghost-positive'
    | 'ghost-black'
    | 'black'
    | 'pink-red';
  padding?: 'sm' | 'md' | 'lg' | 'xl' | 'xxl' | 'square';
  isDisabled?: boolean;
  isLoading?: boolean;
  loadingText?: string;
}

type Ref = HTMLButtonElement;

export const Button = forwardRef<Ref, ButtonProps>(
  (
    {
      variant = 'default',
      className,
      children,
      startIconName,
      endIconName,
      padding = 'md',
      isDisabled = false,
      isLoading = false,
      loadingText = 'Loading',
      ...props
    },
    ref,
  ) => (
    <RadixButton
      className={clsx(className, styles[variant], styles[padding], styles.root)}
      ref={ref}
      disabled={isDisabled || isLoading}
      {...props}
    >
      {isLoading ? (
        <>
          <Icon name="Loader" className="animate-spin" />
          {loadingText}
        </>
      ) : (
        <>
          {startIconName && <Icon name={startIconName} />}
          <Text size="3">{children}</Text>
          {endIconName && (
            <Icon
              name={endIconName}
              className={clsx({
                'ml-auto': variant === 'dropdown',
              })}
            />
          )}
        </>
      )}
    </RadixButton>
  ),
);
