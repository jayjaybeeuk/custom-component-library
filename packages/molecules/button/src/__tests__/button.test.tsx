import { render, screen } from '@testing-library/react';
import { Button } from '../button';

describe('Button', () => {
  it('matches snapshot for default variant and medium padding', () => {
    const { asFragment } = render(<Button>Default Button</Button>);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders with specified variant and padding', () => {
    render(
      <Button variant="positive" padding="lg">
        Positive Button
      </Button>,
    );
    const button = screen.getByRole('button', { name: 'Positive Button' });
    expect(button).toBeInTheDocument();
  });

  it('disables the button when isDisabled is true', () => {
    render(<Button isDisabled>Disabled Button</Button>);
    const button = screen.getByRole('button', { name: 'Disabled Button' });
    expect(button).toBeDisabled();
  });

  it('renders with start and end icons', () => {
    render(
      <Button startIconName="Airplay" endIconName="Activity">
        Button with Icons
      </Button>,
    );
    const icons = screen.getAllByTestId('icon');
    expect(icons.length).toBe(2);
  });

  it('renders with loading icon', () => {
    render(<Button isLoading>Loading</Button>);
    const icons = screen.getAllByTestId('icon');
    expect(icons.length).toBe(1);
  });
});
