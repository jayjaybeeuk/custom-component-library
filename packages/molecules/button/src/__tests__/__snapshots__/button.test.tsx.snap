// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button matches snapshot for default variant and medium padding 1`] = `
<DocumentFragment>
  <button
    class="rt-reset rt-BaseButton rt-r-size-2 rt-variant-solid rt-Button default md root"
    data-accent-color=""
  >
    <span
      class="rt-Text rt-r-size-3"
    >
      Default Button
    </span>
  </button>
</DocumentFragment>
`;
